c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+            // sorted list.  If the first item of such a run was pushed onto the stack, we will
+            // catch it here when processing the second item.  Since this case avoids pushing or
+            // committing any items, all remaining items in the run will land here too.
+            if let Some(other) = stack.last().filter(|other| other.span == span) {
+                if other.rw != rw {
+                    // This item has the same span as the previous one, but wants to perform a
+                    // different rewrite.
+                    errs.push((span, rw, RewriteError::Conflict));
+                }
+                continue;
+            }
+
+            // Commit all rewrites that come strictly before `span`.  This leaves only nodes that
+            // at least partially overlap the current item.  These should normally be ancestors of
+            // the current item on the stack, but it's also possible that the current item
+            // erroneously partially overlaps a previous item.
+            while stack.last().map_or(false, |rt| rt.span.hi() <= lo) {
+                let rt = stack.pop().unwrap();
+                commit(&mut stack, &mut out, rt);
+            }
+
+            // Check for error cases.
+            let siblings = if let Some(parent) = stack.last() {
+                debug_assert!(parent.span.overlaps(span));
+                if !parent.span.contains(span) {
+                    // `span` overlaps `parent.span`, but isn't contained within it.
+                    errs.push((span, rw, RewriteError::PartialOverlap(parent.span)));
+                    continue;
+                }
+                &parent.children
+            } else {
+                &out
+            };
+
+            if let Some(prev) = siblings.last() {
+                debug_assert!(prev.span.lo() <= span.lo());
+                if prev.span.overlaps(span) {
+                    errs.push((span, rw, RewriteError::PartialOverlap(prev.span)));
+                    continue;
+                }
+            }
+
+            // TODO: check that all children are emitted somewhere in the `Rewrite`
+            // (e.g. given the rewrite `f(x + y) -> x`, we can't also have a rewrite on `x + y`,
+            // because that expression is discarded as part of the outer rewrite)
+            //
+            // Specifically: if there is a parent, check that either the current item's span is
+            // contained in the parent span and the parent's rewrite contains `Rewrite::Identity`,
+            // or the current item's span is contained in the span of some `Rewrite::Sub` in
+            // the parent rewrite.  If this doesn't hold, then produce `RewriteError::Discarded`.
+
+            // Push a new node onto the stack.
+            stack.push(RewriteTree {
+                span,
+                rw,
+                children: Vec::new(),
+            });
+        }
+
+        while let Some(rt) = stack.pop() {
```

Comment:
i think this deserves a comment similar to above where you explain that you're committing all rewrites that come before a particular span

Reply:
Added a comment.  This is slightly different from the previous loop since it commits everything left over on the stack instead of only the ones that come before a certain span.

Reply:
Right, I meant similar in the sense of describing why it commits a bunch of things. The comment LGTM